src/managers/OnOfframpManager.sol

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity 0.8.28;
+
+import {CastLib} from "src/misc/libraries/CastLib.sol";
+import {IERC165} from "src/misc/interfaces/IERC165.sol";
+import {SafeTransferLib} from "src/misc/libraries/SafeTransferLib.sol";
+
+import {PoolId} from "src/common/types/PoolId.sol";
+import {ShareClassId} from "src/common/types/ShareClassId.sol";
+import {UpdateContractType, UpdateContractMessageLib} from "src/spoke/libraries/UpdateContractMessageLib.sol";
+
+import {IBalanceSheet} from "src/spoke/interfaces/IBalanceSheet.sol";
+import {IUpdateContract} from "src/spoke/interfaces/IUpdateContract.sol";
+
+import {IOnOfframpManager} from "src/managers/interfaces/IOnOfframpManager.sol";
+import {IOnOfframpManagerFactory} from "src/managers/interfaces/IOnOfframpManagerFactory.sol";
+import {IDepositManager, IWithdrawManager} from "src/managers/interfaces/IBalanceSheetManager.sol";
+
+/// @title  OnOfframpManager
+/// @notice Balance sheet manager for depositing and withdrawing ERC20 assets.
+///         - Onramping is permissionless: once an asset is allowed to be onramped and ERC20 assets have been
+///           transferred to the manager, anyone can trigger the balance sheet deposit.
+///         - Offramping is permissioned: only predefined relayers can trigger withdrawals to predefined
+///           offramp accounts.
+contract OnOfframpManager is IOnOfframpManager {
+    using CastLib for *;
+
+    PoolId public immutable poolId;
+    address public immutable spoke;
+    ShareClassId public immutable scId;
+    IBalanceSheet public immutable balanceSheet;
+
+    mapping(address asset => bool) public onramp;
+    mapping(address asset => mapping(address relayer => address receiver)) public offramp;
+
+    constructor(PoolId poolId_, ShareClassId scId_, address spoke_, IBalanceSheet balanceSheet_) {
+        poolId = poolId_;
+        scId = scId_;
+        spoke = spoke_;
+        balanceSheet = balanceSheet_;
+    }
+
+    //----------------------------------------------------------------------------------------------
+    // Owner actions
+    //----------------------------------------------------------------------------------------------
+
+    /// @inheritdoc IUpdateContract
+    function update(PoolId poolId_, ShareClassId, /* scId */ bytes calldata payload) external {
+        require(poolId == poolId_, InvalidPoolId());
+        require(msg.sender == spoke, NotAuthorized());
+
+        uint8 kind = uint8(UpdateContractMessageLib.updateContractType(payload));
+
+        if (kind == uint8(UpdateContractType.UpdateAddress)) {
+            UpdateContractMessageLib.UpdateContractUpdateAddress memory m =
+                UpdateContractMessageLib.deserializeUpdateContractUpdateAddress(payload);
+
+            if (m.kind == "onramp") {
```

Comment:
QUESTION

Would remove completely. If we make this permissionless we should just allow to onramp any kind of asset permissionlessly. Or what is your worry here?

Reply:
I think one potential issue would be that you could spam 1000s of assets into a balance sheet of a pool, and make the submission of queued assets until it's a snapshot really expensive.

Reply:
Generally seems like something that the pool manager should still control.

Reply:
Makes sense.